### ISSUE TITLE
CD-ROMs are mounted with SATA, instead of IDE controller

### ISSUE THREAD

**Issue by ScriptXaaS** [jetbrains-infra/packer-builder-vsphere#74]
Build Windows VM from ISO on vSphere 6.5 works fine.  However, it fails on vSphere 6.0 for two reasons:
1. CD-ROMs are mounted with SATA, instead of IDE controller.
2. First CD-ROM always showed up a NOT connected.

**Comment by casey-robertson** [comment]
Not my experience.  I've been building against vSphere 6.0 all week.  Can you post your Packer file?

**Comment by mkuzmin** [comment]
These are two independent problems, and it would be easier for both of us to work on them in two separate issues.

Please open a new issue for the second problem. So far I don't think it's caused by vSphere version.

**Comment by mkuzmin** [comment]
I've recently switched the connection from IDE to SATA for macOS compatibility. 

Is there any particular reason you need to have it configured differently?

**Comment by ScriptXaaS** [comment]
I vaguely remebered a previous Windows version may need the ISO  on IDE.  Don't have a particular reason to set it to IDE, just trying different settings to get it working.   Even tried hardcoding the following with no avail:
"configuration_paramaters": {
   "sata0:0.startConnected" : "TRUE"
}

Here is the JSON template:
{
  "variables":{
    "username": "",
    "password": "",
  },

  "builders": [
    {
      "type": "vsphere-iso",

      "vcenter_server": "vc.vsphere.local",
      "cluster": "VDI",
      "insecure_connection": "true",
      "username": "{{user `username`}}",
      "password": "{{user `password`}}",

      "folder": "Test",
      "vm_name":  "pacWin101709-01",
      "convert_to_template": "true",
      "guest_os_type": "windows9_64Guest",

      "communicator": "winrm",
      "winrm_username": "{{user `winrm_username`}}",
      "winrm_password": "{{user `winrm_password`}}",

      "CPUs":             2,
      "RAM":              4096,
      "RAM_reserve_all": true,

      "disk_controller_type":  "pvscsi",
      "disk_size":        40,
      "disk_thin_provisioned": true,

      "network": "VM network",
      "network_card": "vmxnet3",
      "datastore": "datastore1",
      "iso_paths": [
        "[datastore1] ISO/WIN_10_ENT_PRO_EDU_1709_64BIT.iso",
        "[datastore1] ISO/VMware Tools/10.2.0/windows.iso"
      ],

      "floppy_files": [ "win10setup/" ],
      "floppy_img_path": "[datastore1] ISO/VMware Tools/10.2.0/pvscsi-Windows8.flp",
      "boot_order": "disk,cdrom",
      "configuration_parameters":{
        "cpuid.coresPerSocket": "2"

      }
    }
  ],

  "provisioners": [
    {
      "type": "windows-shell",
      "inline": ["dir c:\\"]
    }
  ]
}

**Comment by chris-david-taylor** [comment]
Hi @mkuzmin ,

I can confirm that I'm also having this issue.  Curiously, I just build from master without your commit in place and it's still trying to create instances with SATA cdrom devices.

I reversed all 3 files but strictly I think I only need to change driver/vm.go   

If I have enough time in the next few days I might try adding an option in the builder for IDE/SATA.
Thanks,
Chris

**Comment by ScriptXaaS** [comment]
Hi @mkuzmin,
Just try to build from a Ubuntu template of 16.04 on vSphere 6.5, it appears to have the same problem:  The first SATA CD-ROM will always be NOT connected.  In contrast, building from a Win10 1709 template on vSphere 6.5, the first SATA CD-ROM are always connected.   Here the template I use for Ubuntu:

{
    "variables": {
        "username": "",
        "password": "",
        "ssh_username": "",
        "ssh_password": ""
    },
    
    "builders": [
        {
            "type": "vsphere-iso",
            "vcenter_server": "vc.vsphere.local",
            "username": "{{ user `username` }}",
            "password": "{{ user `password` }}",
            "insecure_connection": "true",
            "vm_name": "pacUbuntu_1604",
            "cluster": "MgmtCluster",
            "host": "esxi-01.vsphere.local",
            "guest_os_type": "ubuntu64Guest",
            "ssh_username": "{{ user `ssh_username` }}",
            "ssh_password": "{{ user `ssh_password` }}",
            
            "CPUs": 2,
            "RAM": 4096,
            "RAM_reserve_all": true,
            "disk_controller_type": "pvscsi",
            "disk_size": 40,
            "disk_thin_provisioned": true,
            "network_card": "vmxnet3",
            "network": "VM Network",
            "datastore": "ses3.esxi.iscsi01",
            "iso_paths": "[ses3.esxi.iscsi01] ISO/ubuntu-16.04.4-server-amd64.iso",
            
           "floppy_files": [
                "{{template_dir}}/preseed.cfg"
            ],
            "boot_command": [
                "<enter><wait><f6><wait><esc><wait>",
                "<bs><bs><bs><bs><bs><bs><bs><bs><bs><bs>",
                "<bs><bs><bs><bs><bs><bs><bs><bs><bs><bs>",
                "<bs><bs><bs><bs><bs><bs><bs><bs><bs><bs>",
                "<bs><bs><bs><bs><bs><bs><bs><bs><bs><bs>",
                "<bs><bs><bs><bs><bs><bs><bs><bs><bs><bs>",
                "<bs><bs><bs><bs><bs><bs><bs><bs><bs><bs>",
                "<bs><bs><bs><bs><bs><bs><bs><bs><bs><bs>",
                "<bs><bs><bs><bs><bs><bs><bs><bs><bs><bs>",
                "<bs><bs><bs>",
                "/install/vmlinuz",
                " initrd=/install/initrd.gz",
                " priority=critical",
                " locale=en_US",
                " file=/media/preseed.cfg",
                "<enter>"
            ],
            "boot_order": "disk,cdrom"
        }
    ],
    "provisioners": [
        {
            "type": "shell",
            "inline": [
                "ls /"
            ]
        }
    ]
}

**Comment by mkuzmin** [comment]
I still cannot reproduce the issue. Let's clarify details.
Is CD-ROM device created?
Is ISO file mounted in VM settings?
Open interactive vSphere console and try to connect the drive manually. Does it produce any errors?

**Comment by chris-david-taylor** [comment]
I had a quick go trying to reproduce this @mkuzmin and it works for me.  I'm also using a number of Windows and *nixes without problem (when I said it didn't work previously, Admins changed my permissions without telling me).  I can't help thinking it might be something simpler like a corrupted ISO file. a bit of Googling reveal that a lot of users of 16.04 are having issues with SATA.  
It seems unlikely to me that its a problem that can be fixed here.

**Comment by ScriptXaaS** [comment]
I just verified on 6.5 with both Win10 v1709 and Ubuntu 1604. Both works now.  The original failure on Ubuntu 16.04  is due to wrong datastore path.  When the datastore path is incorrect, CD-ROM device will be created, but showed up NOT connected.   With the latest vCenter 6.5 u1g version, I clicked on connect checkbox and it informed that the datastore path being incorrect.  I vaguely remember in the previous vCenter 6.5 version (using HTML5 client), I could not click on connected checkbox and there was no error message.  Thanks for looking into this.
